In [1]:
import numpy as np
from scipy import stats
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

# 과제 1 정규분포의 확률계산(43pp)

# 1) 1리터를 가지고 9km 이상 15km 이하를 달릴 확률은?

In [14]:
start = 9 
mu = 12 #자동자차의 1리터당 평균 주행거리
sigma = 3 #표준편차 
x1 =15 
prob1 = stats.norm.cdf(x1, mu, sigma)
print("{0}km 이상 {1}km 이하를 달릴 확률은 {2:.3f}에 해당함".format(start, x1, (prob1-0.5)*2))

9km 이상 15km 이하를 달릴 확률은 0.683에 해당함


1) 1리터를 가지고 9km 이상 15km 이하를 달릴 확률은 0.683에 해당합니다. 이는 표준정규분포의 -1시그마 <= 평균 <= 1시그마 사이의 확률을 나타냅니다. 이는 평균에서 1시그마의 확률 값에 * 2를 한 것과 동일 합니다. 1시그마는 0.341 이며 이를 2배하면 대략 0.682~3 이 나옵니다. 


# 정규성 검증을 하고 검증해야 합니다.

# 2) 1리터를 가지고 15km 이상을 달릴 확률은?

In [22]:
mu = 12 #자동자차의 1리터당 평균 주행거리
sigma = 3 #표준편차 
x1 =15 
prob2 = stats.norm.cdf(x1, mu, sigma) # -무한대에서 15km까지의 cdf 값
#print("{0}km 까지의 cdf의 값은 {1:.3f}에 해당함".format(x1, prob2))
print("{0}km 이상을 달릴 확률은 {1:.3f}에 해당함".format(x1, (1-prob2)))

15km 이상을 달릴 확률은 0.159에 해당함


2) 15까지의 누적분포값은 0.841이므로 1리터를 가지고 15km 이상을 달릴 확률은 1에서 해당 누적분포값(0.841)을 제외하면 되기 때문에 0.159입니다. 

# 과제2. 포아송 분포(78pp)

In [37]:
mu = 3
prob_2 = stats.poisson.pmf(2, mu) #1) 2명의 손님이 도착
prob_1 = stats.poisson.cdf(1, mu) #2) 1명 이하의 손님이 도착
prob_3 = 1 - (stats.poisson.cdf(2, mu)) #3) 3명 이상의 손님이 도착
print("2명의 손님이 도착할 확률은 {0:.3f}이며,\n 1명 이하의 손님이 도착할 확률은 {1:.3f}이고,\n 3명 이상의 손님이 도착할 확률은 총합 1에서 2명 이하의 손님이 도착할 확률을 뺀 {2:.3f}이다.\n".format(prob_2, prob_1, prob_3))

2명의 손님이 도착할 확률은 0.224이며,
 1명 이하의 손님이 도착할 확률은 0.199이고,
 3명 이상의 손님이 도착할 확률은 총합 1에서 2명 이하의 손님이 도착할 확률을 뺀 0.577이다.



### 포아송 분포를 활용하며 2명의 손님이 도착할 확률은 이산확률 분포인 pmf(Probability mass function)을 활용하였고, 1명 이하와 3명 이상 손님이 도착할 확률은 cdf(Cumulate Density function)을 활용하여서 계산하였습니다. 

# 과제3. 정규성 검정(96pp)

## 1. 가설 설정

1)가설 설정
- 귀무 가설: MAT_A는 정규성을 따른다.
- 대립 가설: MAT_A는 정규성을 따르지 않는다.

2)가설 설정
- 귀무 가설: MAT_B는 정규성을 따른다.
- 대립 가설: MAT_B는 정규성을 따르지 않는다.

## 2. 데이터 분석

In [42]:
# 데이터 불러오기
ds = pd.read_csv("C:/Users/hyang/Downloads/1_datasets/2sample_t_test.csv")
# 결측치 제거
ds = ds.dropna()
#변수 할당
data1 = ds['Mat_A']
data2 = ds['Mat_B']
# 정규성 검정
from scipy.stats import shapiro
shapiro(data1)
shapiro(data2)
#statistic

ShapiroResult(statistic=0.9729756116867065, pvalue=0.9169799089431763)

## 3. 결과 확인

In [44]:
statistic1, p1=stats.shapiro(data1)
statistic2, p2=stats.shapiro(data2)
print("Shapiro-Wilk Test: statistic={:1.3f}, p-value={:1.3f}".format(statistic1, p1))
print("Shapiro-Wilk Test: statistic={:1.3f}, p-value={:1.3f}".format(statistic2, p2))

Shapiro-Wilk Test: statistic=0.906, p-value=0.256
Shapiro-Wilk Test: statistic=0.973, p-value=0.917


#### 변수 MAT_A와 MAT_B에 대해 각각 정규성 검사를 한 결과, MAT_A의 통계량은 0.906, p값은 0.256으로 유의수준(0.05)보다 높은 값을 보였다.
#### 또한, 변수 MAT_B의 통계량은 0.973, p값은 0.917으로 95%의 신뢰수준을 사용할 때 유의수준(0.05)보다 높은 값을 보였다.
#### 따라서, 두 변수의 귀무 가설이 모두 채택되어 두 변수는 모두 정규성을 따른다.

# 과제 4. 모평균 추정(99pp)

1. 손으로 직접 계산했을 때의 새로운 품종 감자 20개의 무게의 모표준편차가 3.8으로 알려져 있을 때 95%의 신뢰구간은
x - 1.96 *(3.8/sqrt(20) <= 모 평균 <= x + 1.96 *(3.8/sqrt(20) 
평균은 19.5 +- 1.6654 이기 때문에 95%의 신뢰구간은 17.8346 <= 모평균 <= 21.1654 라고 결과 값이 나온다. 

In [45]:
sample = pd.DataFrame({"sample" : [18, 18, 20, 21, 20, 23, 19, 18, 17, 21, 22, 20, 20, 21, 20, 19, 19, 18, 17, 19]})

In [49]:
sample.describe() #데이터 확인

sample
count  20.00000
mean   19.50000
std     1.60591
min    17.00000
25%    18.00000
50%    19.50000
75%    20.25000
max    23.00000

In [50]:
lower, upper = stats.norm.interval(0.95, loc=np.mean(sample), scale=3.8/np.sqrt(20))
print("신뢰구간 : ({0}, {1})".format(lower.round(4), upper.round(4)))

신뢰구간 : ([17.8346], [21.1654])


### 결과 : 새로운 품종의 감자 무게의 모평균은 95%의 신뢰구간 17.83과 21.16 사이에 위치한다

# 과제5. 모평균 추정(102pp) - 모표준편차를 모르는 경우

In [51]:
product_test = pd.DataFrame({'product': [73, 71, 74, 69, 70, 73, 70, 68, 75, 72, 70, 72, 73, 70, 70, 72, 71, 70, 75, 72]})

In [65]:
product_test.describe() #데이터 확인 평균 71.5 맞음

product
count  20.000000
mean   71.500000
std     1.933091
min    68.000000
25%    70.000000
50%    71.500000
75%    73.000000
max    75.000000

In [61]:
# product_test.isna().count() #결측치 확인

In [63]:
import scipy
lower, upper = stats.t.interval(0.95, len(product_test)-1, loc=np.mean(product_test), scale=scipy.stats.sem(product_test))

In [64]:
print("신뢰구간: ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간: ([70.6], [72.4])


### 결과: 제품 20개의 표본평균 71.50과 표본표준편차 1.993을 활용하여, 95%의 신뢰구간 70.6 ~ 72.4 사이에 새로 개발된 강종의 평균 인장강도가 있다고 추정할 수 있습니다.